odoo/addons/base/models/res_company.py

```diff
@@ -354,6 +368,9 @@ def write(self, values):
                 for fname in sorted(changed):
                     branches[fname] = company[fname]
 
+        if companies_needs_l10n:
+            companies_needs_l10n.install_l10n_modules()
```

Comment:
This is the source of your nightmare, and will never work correctly.
Installing a module requires to reload the registry, i.e., instantiate a **new** `Registry` object.
After the installation, the existing recordset objects belong to a Python class which still refers to the **old** registry, and that's why those recordsets don't have access to the newly created fields, etc.

Therefore, doing anything with existing records after installing a module from business code is always a source of problems.  That's why the installation of modules in `res.config.settings` is the last thing being done when applying the new config, for instance.  In other words, installing modules _in the middle_ of regular business code will never work properly.  It's a technical limitation that is very hard to get around.

In your case, creating a company on the fly or modifying a company's country will potentially install a module.  And once that module is installed, resuming and using existing recordsets is a mistake, because their class corresponds to an outdated registry.  I am afraid you will hardly ever get around it.